internal/lib/stats/stats_server_linux.go

```diff
@@ -119,6 +150,100 @@ func (ss *StatsServer) populateNetworkUsage(stats *types.PodSandboxStats, sb *sa
 	})
 }
 
+// metricsForPodSandbox is an internal, non-locking version of MetricsForPodSandbox
+// that returns (and occasionally gathers) the metrics for the given sandbox.
+// Note: caller must hold the lock on the StatsServer
+func (ss *StatsServer) metricsForPodSandbox(sb *sandbox.Sandbox) *SandboxMetrics {
+	if ss.collectionPeriod == 0 {
+		return ss.updatePodSandboxMetrics(sb)
+	}
+	sboxMetrics, ok := ss.sboxMetrics[sb.ID()]
+	if ok {
+		return sboxMetrics
+	}
+	// Cache miss, try again
+	return ss.updatePodSandboxMetrics(sb)
+}
+
+// updatePodSandboxMetrics updates the sandbox metrics for the given sandbox.
+// If the sandbox is not found, it creates a new entry in the map.
+// Note: caller must hold the lock on the StatsServer.
+func (ss *StatsServer) updatePodSandboxMetrics(sb *sandbox.Sandbox) *SandboxMetrics {
+	if sb == nil {
+		return nil
+	}
+	sm, exists := ss.sboxMetrics[sb.ID()]
+	if !exists {
+		sm = NewSandboxMetrics(sb)
+	}
+	// network metrics are collected at pod level
+	if slices.Contains(ss.Config().IncludedPodMetrics, "network") {
+		podMetrics := ss.GenerateNetworkMetrics(sb)
+		sm.metric.Metrics = podMetrics
+	}
+
+	containersList := sb.Containers().List()
+	containerMetrics := make([]*types.ContainerMetrics, 0, len(containersList))
+	for _, c := range containersList {
+		// Skip if the container is stopped
+		if c.StateNoLock().Status == oci.ContainerStateStopped {
+			continue
+		}
+		cMetrics := ss.GenerateSandboxContainerMetrics(sb, c, sm)
+		containerMetrics = append(containerMetrics, cMetrics)
+	}
+	sm.metric.ContainerMetrics = containerMetrics
+	ss.sboxMetrics[sb.ID()] = sm // Update the entry in the map
+	return sm
+}
+
+// GenerateSandboxContainerMetrics generates a list of metrics for the specified sandbox
+// containers by collecting metrics from the cgroup based on the includedMetrics
+// except for network metrics which is collected at pod level.
+func (ss *StatsServer) GenerateSandboxContainerMetrics(sb *sandbox.Sandbox, c *oci.Container, sm *SandboxMetrics) *types.ContainerMetrics {
+	cgstats, err := ss.Runtime().ContainerStats(context.TODO(), c, sb.CgroupParent())
+	if err != nil || cgstats == nil {
+		logrus.Errorf("Error getting sandbox stats %s: %v", sb.ID(), err)
+		return nil
+	}
+	return ss.containerMetricsFromCgStats(sb, c, cgstats, sm)
+}
+
+func (ss *StatsServer) containerMetricsFromCgStats(sb *sandbox.Sandbox, c *oci.Container, cgstats *cgmgr.CgroupStats, sandboxMetrics *SandboxMetrics) *types.ContainerMetrics {
+	var metrics []*types.Metric
+	for _, m := range ss.Config().IncludedPodMetrics {
+		switch m {
+		case "cpu":
```

Comment:
note to self: verify how the kubelet correlates the metrics descriptors  to the metrics we return. It's possible we may need to be very dilligent about the ordering of which metrics we're including

Reply:
actually, if we use the same list here and [here](https://github.com/cri-o/cri-o/pull/7796/files#diff-dbba0afd4b7f28a68d1393c2e4bd53d9e92a38b691b6e3bba1b8b87d9a709699R10) then we should be okay, as long as the actual metrics underneath are also in the same order

Reply:
wait but if we put them in a [map](https://github.com/cri-o/cri-o/pull/7796/files#diff-dbba0afd4b7f28a68d1393c2e4bd53d9e92a38b691b6e3bba1b8b87d9a709699R11) and then transfer that to a slice, we'll lose the ordering...